### PR TITLE
Code Review

### PR DESCRIPTION
**Pull request by capoferro**
This is the beginnings of a combat simulator for FFXIV. It's basically a reimplementation of the game's combat system with the intention of running many iterations to find optimal rotations/gear for various jobs. Right now, everything is just hard-coded... some day it'll take rotation and stat input as data.

I wanted to learn Bevy and rust, so this was an interesting project to motivate me. I'm a complete novice at both.

### CODE REVIEW

File: src/actor.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    &'static Actions,
+    &'static Rotation,
+    &'static mut RecastExpirations,
+    &'static mut Damage,
+    &'static mut StatusEffects,
+    &'static mut Stats,
+    &'static mut ActiveCombos,
+);
+
+pub type QueryActor<'a> = Query<'a, ActorBundle>;
```
Comment (alice-i-cecile) [inline src/actor.rs]: This pattern feels very OOP. Querying for all components of an entity is pretty much always an antipattern where I've seen it.

File: src/actor.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+use apply::Apply;
+use bevy_ecs::prelude::{Entity, Query};
+use calc::lookup::Job;
+use damage::Damage;
+use recast_expirations::RecastExpirations;
+use rotation::Rotation;
+use stat::Stats;
+use status_effect::StatusEffects;
+
+pub type ActorBundle = (
```
Comment (alice-i-cecile) [inline src/actor.rs]: It's  a bit odd to call this `ActorBundle` when it doesn't have the `Bundle` trait.

File: src/actor/action.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+#[derive(Default, Clone)]
+pub struct Action {
+    pub id: u32,
+    pub name: &'static str,
+    pub ogcd: bool,
+    pub results: Vec<Arc<dyn Apply + Send + Sync>>,
+}
+
+impl Action {
+    pub fn perform(&self, sim: &SimState, query: &mut QueryActor, source: Entity, target: Entity) {
```
Comment (alice-i-cecile) [inline src/actor/action.rs:34]: This is feeling pretty OOP with a big monolithic state. For your game, the performance almost certainly doesn't matter. But a more modular design may be easier to extend test and reason about.

File: src/actor/action.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+#[derive(Default)]
+pub struct Actions(HashMap<u32, Action>);
+
+impl Actions {
+    pub fn add(&mut self, action: Action) {
+        self.0.insert(action.id, action);
+    }
+
+    delegate! {
```
Comment (alice-i-cecile) [inline src/actor/action.rs:17]: The delegate macro is really cool here! That's a neat pattern I've never seen in Bevy before.

File: src/actor/action.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            pub fn get(&self, id: &u32) -> Option<&Action>;
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Action {
+    pub id: u32,
+    pub name: &'static str,
+    pub ogcd: bool,
```
Comment (alice-i-cecile) [inline src/actor/action.rs:29]: On global cooldown? Using a newtype pattern for bools like this can often be better for communicating intent and avoiding dumb bugs.

File: src/actor/active_combos.rs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+use delegate::delegate;
+use std::collections::HashSet;
+
+#[derive(Default)]
+pub struct ActiveCombos(HashSet<u32>);
+
+impl ActiveCombos {
```
Comment (alice-i-cecile) [inline src/actor/active_combos.rs:7]: The intent of this code is very unclear at first glance, I'm not sure that the macros here are net useful, especially since they'll make the code harder to expand and revise.

File: src/actor/apply.rs [hunk 7]
```diff
@@ -0,0 +1,155 @@
+use super::calc;
+use super::status_effect::status::{SetCombo, Status, StatusFlag, StatusFlags};
+use super::status_effect::StatusEffect;
+use super::{ActiveCombos, QueryActor};
+use crate::sim::{SimState, SimTime};
+use bevy_ecs::prelude::Entity;
+use std::sync::Arc;
+
+pub trait Apply {
+    fn apply(&self, sim: &SimState, query: &mut QueryActor, source: Entity, target: Entity);
```
Comment (alice-i-cecile) [inline src/actor/apply.rs:10]: Decoupling "what should be done" from "actually doing the thing" using events is a very common and powerful pattern for breaking apart the monolithic state you see in `SimState` and `QueryActor`.

File: src/actor/apply.rs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+impl Apply for DoDirectDamage {
+    fn apply(&self, sim: &SimState, query: &mut QueryActor, source: Entity, target: Entity) {
+        let calculated_damage;
+        if let Ok((_, job, _, _, _, _, mut status_effects, stats, mut active_combos)) =
+            query.get_mut(source)
+        {
+            let mut potency = self.potency;
+            if self.consume_combo(&mut active_combos) {
+                potency = match self.combo_potency {
```
Comment (alice-i-cecile) [inline src/actor/apply.rs]: `.expect` is a much clearer way to implement this basic pattern.

File: src/actor/apply.rs [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    fn apply(&self, sim: &SimState, query: &mut QueryActor, source: Entity, _target: Entity) {
+        if let Ok((_, _, _, _, mut recast_expirations, _, _, _, _)) = query.get_mut(source) {
+            recast_expirations.set(self.action_id, sim.now() + self.duration);
+        }
+    }
+}
+
+pub struct GiveStatusEffect {
+    pub status: Status,
+    pub target_source: bool,
```
Comment (alice-i-cecile) [inline src/actor/apply.rs:76]: This is a bit surprising that it's a bool?

File: src/actor/apply.rs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    fn potency() {
+        let ddd = DoDirectDamage {
+            potency: 100,
+            combo_potency: Some(200),
+            combo_action_id: Some(1),
+            ..Default::default()
+        };
+        let mut active_combos = ActiveCombos::default();
+        active_combos.add_action(2);
+        assert_eq!(false, ddd.consume_combo(&mut active_combos));
```
Comment (alice-i-cecile) [inline src/actor/apply.rs:146]: Returning true and false from this method is quite unclear, as seen in the tests.

File: src/actor/apply.rs [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+    #[test]
+    fn potency() {
+        let ddd = DoDirectDamage {
+            potency: 100,
+            combo_potency: Some(200),
+            combo_action_id: Some(1),
+            ..Default::default()
+        };
+        let mut active_combos = ActiveCombos::default();
+        active_combos.add_action(2);
```
Comment (alice-i-cecile) [inline src/actor/apply.rs:145]: Using integers to identify your actions seems hard to understand.

File: src/actor/calc.rs [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+// https://www.akhmorning.com/allagan-studies/how-to-be-a-math-wizard/shadowbringers/damage-and-healing/#direct-damage-d
+pub fn direct_damage(
+    sim: &SimState,
+    potency: i64,
+    job: lookup::Job,
+    stats: &Stats,
+    attack_type: AttackType,
+    multipliers: Vec<f64>,
+) -> i64 {
+    let fatk = attack_power(job, stats.get(Stat::AttackPower));
```
Comment (alice-i-cecile) [inline src/actor/calc.rs:28]: The variable names here are particularly terse, and hard to follow as someone who's not terribly familiar with FFXIV.

File: src/actor/calc.rs [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+    // D = ⌊ D3 × rand[95,105] ⌋ /100 ⌋
+    let d = d3 * sim.rng.random_from_range(95, 106) / 100;
+
+    // ⌊ ⌊ D × buff_1 ⌋ × buff_2 ⌋
+    multipliers
+        .iter()
+        .fold(d as f64, |total, multiplier| floor(total * *multiplier, 0)) as i64
+}
+
+// Level 80 F(AP)
```
Comment (alice-i-cecile) [inline src/actor/calc.rs]: Comments like this should be doc comments whenever you can manage it (using `///`), so then they're more easily accessible to end users and in IDE highlights.

File: src/actor/calc.rs [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+        let sim = SimState::new(FakeRng {
+            random_value: 0.5,
+            random_from_range_value: 100,
+        });
+
+        assert_eq!(true, is_crit(&sim, 0, Some::<&i64>(&51)));
+    }
+
+    /*
+    fast blade potency 200
```
Comment (alice-i-cecile) [inline src/actor/calc.rs]: The purpose of this comment block is very unclear.

File: src/actor/calc/lookup.rs [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            Job::GLA | Job::MRD | Job::PLD | Job::WAR | Job::DRK | Job::GNB => Stat::Strength,
+            Job::THM | Job::ACN | Job::BLM | Job::SMN | Job::RDM | Job::BLU => Stat::Intelligence,
+            Job::CNJ | Job::WHM | Job::SCH | Job::AST => Stat::Mind,
+            _ => panic!("Tried to get primary stat of unknown job: {:?}", self),
+        }
+    }
+
+    pub fn trait_multiplier(&self) -> i64 {
+        match self {
+            Job::ARC | Job::BRD | Job::MCH | Job::DNC => 120,
```
Comment (alice-i-cecile) [inline src/actor/calc/lookup.rs:76]: In general, large look-up tables may be more reasonable to store as data and then load. The current approach is fine however.

File: src/actor/stat.rs [hunk 16]
```diff
@@ -0,0 +1,115 @@
+use std::collections::HashMap;
+#[allow(dead_code)]
+#[derive(PartialEq, Eq, Hash, Copy, Clone)]
+pub enum Stat {
```
Comment (alice-i-cecile) [inline src/actor/stat.rs:4]: This definitely feels like it should be a trait instead.

File: src/main.rs [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+        Rotation::default(),
+        RecastExpirations::default(),
+        Damage::default(),
+        StatusEffects::default(),
+        Stats::default(),
+        ActiveCombos::default(),
+    ));
+}
+
+fn tick(mut sim_state_query: Query<&mut SimState>) {
```
Comment (alice-i-cecile) [inline src/main.rs:129]: Have you seen Bevy `State` abstraction that attempts to do something very similar?

File: src/actor/action.rs (hunk 5 above)
Comment (capoferro) [reply src/actor/action.rs:29]: It means "off global cooldown"... you're right that outside FFXIV, it's not clear, though within the community it's common to refer to a job's "oGCDs" as it's a core part of rotation development. Definitely something that deserves a comment.

File: src/actor/action.rs (hunk 3 above)
Comment (capoferro) [reply src/actor/action.rs:34]: Yeah, I'll follow up in the discord thread as I feel like this is a more involved discussion... I'm only doing this cause it works, and it's definitely showing signs of maintenance problems. I'd love to find a better way to provide generic actions.

File: src/actor/apply.rs (hunk 9 above)
Comment (capoferro) [reply src/actor/apply.rs:76]: It's supposed to say "target the source instead of the target". The receiver of the status effect is set on 83 depending on this bool. There's probably a better name for it, as I can see what you mean that it's unclear in isolation.

File: src/actor/apply.rs (hunk 11 above)
Comment (capoferro) [reply src/actor/apply.rs:145]: It should be some unique identifier that is static for each action.... probably from some global ID generator of some sort. Maybe they should even be an Entity?

I agree that hardcoding it is fragile... I am thinking that when this is read from some source data, the ID will be propagated through the objects automatically and not need to be maintained like it is now.

File: src/actor/apply.rs (hunk 10 above)
Comment (capoferro) [reply src/actor/apply.rs:146]: the return is supposed to indicate whether it consumed a combo or not (eg. the combo existed to be consumed, and so `true` that it was)... that is a bit unintuitive though. Will have to puzzle over how to represent that better.

File: src/actor/calc.rs (hunk 12 above)
Comment (capoferro) [reply src/actor/calc.rs:28]: They're intended to be as close to the resource documentation as possible (as seen in the inline comments containing the copied formulas) to make it easy to verify that the translation of the formulas are correct.

`fatk` = `f(ATK)`  
`fdet` = `f(DET)`

Otherwise, I agree completely... fatk is completely meaningless outside of that context.

File: src/actor/calc.rs (hunk 14 above)
Comment (capoferro) [reply src/actor/calc.rs]: Oops, it's just a combat log. Should be deleted.

File: src/main.rs [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+    let mut sim_state = sim_state_query
+        .single_mut()
+        .expect("There should always be exactly one sim state.");
+
+    let now = sim_state.tick();
+    // TODO: Temporary. Sim 15s for now while in dev.
+    println!("============ Tick: {}", now);
+    if now >= 15000 {
+        std::process::exit(0);
```
Comment (IceSentry) [inline src/main.rs:138]: You might want to use the AppExit event instead of exiting directly 
https://github.com/bevyengine/bevy/blob/6aedb2500aedaab449a94e7c972d800c6f5fc56b/examples/ecs/ecs_guide.rs#L125

File: src/main.rs (hunk 17 above)
Comment (capoferro) [reply src/main.rs:129]: Do you mean this? https://bevy-cheatbook.github.io/programming/states.html

Comment (alice-i-cecile) [reply src/main.rs:129]: Yes, precisely. There's also an example here: https://github.com/bevyengine/bevy/blob/latest/examples/ecs/state.rs

File: src/actor/action.rs (hunk 5 above)
Comment (alice-i-cecile) [reply src/actor/action.rs:29]: Yeah, I used to play WoW reasonably seriously so I could guess :)

File: src/actor/apply.rs (hunk 11 above)
Comment (alice-i-cecile) [reply src/actor/apply.rs:145]: Could be hashed from the action's name perhaps?

File: src/actor/calc.rs (hunk 12 above)
Comment (alice-i-cecile) [reply src/actor/calc.rs:28]: Makes sense; I've seen the same thing in scientific programming.
